query-algebrizer/src/clauses/where_fn.rs

```diff
@@ -0,0 +1,486 @@
+// Copyright 2016 Mozilla
+//
+// Licensed under the Apache License, Version 2.0 (the "License"); you may not use
+// this file except in compliance with the License. You may obtain a copy of the
+// License at http://www.apache.org/licenses/LICENSE-2.0
+// Unless required by applicable law or agreed to in writing, software distributed
+// under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
+// CONDITIONS OF ANY KIND, either express or implied. See the License for the
+// specific language governing permissions and limitations under the License.
+
+use std::rc::Rc;
+
+use mentat_core::{
+    Schema,
+    SQLValueType,
+    TypedValue,
+    ValueType,
+};
+
+use mentat_query::{
+    Binding,
+    FnArg,
+    NonIntegerConstant,
+    SrcVar,
+    Variable,
+    VariableOrPlaceholder,
+    WhereFn,
+};
+
+use clauses::{
+    ConjoiningClauses,
+    PushComputed,
+};
+
+use errors::{
+    BindingError,
+    Error,
+    ErrorKind,
+    Result,
+};
+
+use super::QualifiedAlias;
+
+use types::{
+    ComputedTable,
+    SourceAlias,
+    ValueTypeSet,
+    VariableColumn,
+};
+
+macro_rules! check_type_set {
+    ($types: expr, $type: path) => { {
+        if !$types.contains($type) {
+            bail!(ErrorKind::InvalidGroundConstant);
+        }
+    } }
+}
+
+/// Conversion of FnArgs to TypedValues.
+impl ConjoiningClauses {
+    /// Convert the provided `FnArg` to a `TypedValue`.
+    /// The conversion depends on, and can fail because of:
+    /// - Existing known types of a variable to which this arg will be bound.
+    /// - Existing bindings of a variable `FnArg`.
+    fn typed_value_from_arg<'s>(&self, schema: &'s Schema, arg: FnArg, known_types: ValueTypeSet) -> Result<TypedValue> {
+        if known_types.is_empty() {
+            bail!(ErrorKind::InvalidGroundConstant);
+        }
+
+        match arg {
+            // This is potentially ambiguous.
+            FnArg::EntidOrInteger(x) => {
+                match (ValueType::Ref.accommodates_integer(x),
+                       known_types.contains(ValueType::Ref),
+                       known_types.contains(ValueType::Long)) {
+                    (true, true, true) => {
+                        // Ambiguous: this arg could be an entid or a long.
+                        // We default to long.
```

Comment:
There are four stages we can apply.

One is during algebrizing, which I've implemented here: if we know a type is impossible (-1 can't be an entid; `:nope` isn't an ident), then we can narrow down to an individual type, and there's no ambiguity.

The second is also during algebrizing: if we know that `?x` will later be used as an attribute, then we don't need to treat it as a keyword. This can be confusing: after all, one could write a query like:

```
[:find … :where [(ground :foo/bar) ?x] [?a ?x 15] [?a :some/keyword ?x]]
```

— there `?x` can be the same EDN value used in two different places as two different "purpose types" (ref and keyword).

This is part-implemented: it uses the current state of the CC. We would need to do more than one pass through the input patterns in order to determine types more generally.

The third is to defer this decision until the latest possible moment. That's not really feasible: there's no "last", because we would need to iteratively converge.

The last is to defer resolution of user input until the query is underway: to generate the appropriate mess of `UNION` expressions to allow the value to pass as either type.

Rather than doing that, I've followed your approach: if you enter `[(ground 10) ?foo]`, then you'll get a `long`… unless somewhere else on the query unambiguously requires `?foo` to be an entity.

This particular bit of code says: the value is in range of being a ref, and nothing else in the query has yet told us that it's one or the other, so we just pick one.